Plugins/GeoModel/src/GeoModelDetectorObjectFactory.cpp

```diff
@@ -0,0 +1,215 @@
+
+    // go through each fpv
+    for (auto &[name, fpv] : qFPV) {
+      PVConstLink physVol{fpv};
+      // if the match lambda returns false skip the rest of the loop
+      if (!matches(name, physVol)) {
+        continue;
+      }
+      convertFpv(name, fpv, cache, gctx);
+    }
+  }
+}
+
+void Acts::GeoModelDetectorObjectFactory::convertSensitive(
+    PVConstLink geoPV, const Acts::Transform3 &transform,
+    std::vector<GeoModelSensitiveSurface> &sensitives) {
+  const GeoLogVol *logVol = geoPV->getLogVol();
+  const GeoShape *shape = logVol->getShape();
+  int shapeId = shape->typeID();
+  std::string name = logVol->getName();
+  std::shared_ptr<const Acts::IGeoShapeConverter> converter =
+      Acts::GeoShapesConverters(shapeId);
+  if (converter == nullptr) {
+    throw std::runtime_error("The converter for " + recType(*shape) +
+                             " is nullptr");
+
+    return;
+  }
+  auto converted = converter->toSensitiveSurface(geoPV, transform);
+  if (converted.ok()) {
+    sensitives.push_back(converted.value());
+    const auto &[el, sf] = converted.value();
+    auto ss = std::get<1>(converted.value());
+
+    ACTS_VERBOSE("(successfully converted: "
+                 << name << " / " << recType(*shape) << " / "
+                 << logVol->getMaterial()->getName() << ")");
+
+    if (!el || !sf) {
+      throw std::runtime_error("The Detector Element or the Surface is nllptr");
+    }
+    return;
+  }
+  ACTS_ERROR(name << " / " << recType(*shape)
+                  << ") could not be converted by any converter");
+}
+
+std::vector<GeoChildNodeWithTrf>
+Acts::GeoModelDetectorObjectFactory::findAllSubVolumes(PVConstLink vol) {
+  std::vector<GeoChildNodeWithTrf> subvolumes = getChildrenWithRef(vol, false);
+  std::vector<GeoChildNodeWithTrf> sensitives;
+  for (auto subvolume : subvolumes) {
+    if (matches(subvolume.nodeName, subvolume.volume)) {
+      sensitives.push_back(subvolume);
+    }
+    std::vector<GeoChildNodeWithTrf> senssubsubvolumes =
+        findAllSubVolumes(subvolume.volume);
+    std::transform(std::make_move_iterator(senssubsubvolumes.begin()),
+                   std::make_move_iterator(senssubsubvolumes.end()),
+                   std::back_inserter(sensitives),
+                   [&subvolume](GeoChildNodeWithTrf &&volume) {
+                     volume.transform = subvolume.transform * volume.transform;
+                     return volume;
+                   });
+  }
+  return sensitives;
+}
+bool Acts::GeoModelDetectorObjectFactory::convertBox(std::string name) {
+  return (name.find(m_cfg.convertBox) != std::string::npos);
+}
+void Acts::GeoModelDetectorObjectFactory::convertFpv(
+    std::string name, GeoFullPhysVol *fpv, Cache &cache,
+    const GeometryContext &gctx) {
+  PVConstLink physVol{fpv};
+
+  // get children
+  std::vector<GeoChildNodeWithTrf> subvolumes =
+      getChildrenWithRef(physVol, false);
+  if (subvolumes.size() > 0) {
```

Comment:
ok so this means that we convert always the children as long as they exist right? Is this working for all the cases? (MDTs, RPCs etc and also the ID) Because I was thinking of also checking with the matching function for the children..

Reply:
The routine for finding the children is called whenever they exist, but if none of the children names and materials match the config.nameList and config.materialList then nothing will be converted.